controllers/integration/integration_adapter.go

```diff
@@ -0,0 +1,310 @@
+	if err != nil {
+		return results.RequeueWithError(err)
+	} else if existingApplicationSnapshot == nil {
+		return results.RequeueAfter(time.Second*5, err)
+	}
+
+	releaseLinks, err := a.getAllApplicationReleaseLinks()
+	if err != nil {
+		a.logger.Error(err, "Failed to get all ReleaseLinks",
+			"Application.Name", a.application.Name,
+			"Application.Namespace", a.application.Namespace)
+		return results.RequeueOnErrorOrStop(a.updateStatus())
+	}
+
+	releases, err := a.getAllApplicationSnapshotReleases(existingApplicationSnapshot)
+	if err != nil {
+		a.logger.Error(err, "Failed to get all Releases",
+			"ApplicationSnapshot.Name", existingApplicationSnapshot.Name,
+			"ApplicationSnapshot.Namespace", existingApplicationSnapshot.Namespace)
+		return results.RequeueOnErrorOrStop(a.updateStatus())
+	}
+	for _, releaseLink := range *releaseLinks {
+		var existingRelease *releasev1alpha1.Release = nil
+		for _, snapshotRelease := range *releases {
+			if snapshotRelease.Spec.ReleaseLink == releaseLink.Name {
+				existingRelease = &snapshotRelease
+			}
+		}
+		if existingRelease != nil {
+			a.logger.Info("Found existing Release",
+				"ApplicationSnapshot.Name", existingApplicationSnapshot.Name,
+				"ReleaseLink.Name", releaseLink.Name,
+				"Release.Name", existingRelease.Name)
+		} else {
+			newRelease := a.createReleaseForReleaseLink(&releaseLink, existingApplicationSnapshot)
+			err = a.client.Create(a.context, newRelease)
+			if err != nil {
+				a.logger.Error(err, "Failed to create new Release",
+					"ApplicationSnapshot.Name", existingApplicationSnapshot.Name,
+					"ApplicationSnapshot.Namespace", existingApplicationSnapshot.Namespace,
+					"ReleaseLink.Name", releaseLink.Name)
+				return results.RequeueOnErrorOrStop(a.updateStatus())
+			}
+			a.logger.Info("Created new Release",
+				"Application.Name", a.application.Name,
+				"ReleaseLink.Name", releaseLink.Name,
+				"Release.Name", newRelease.Name)
+		}
+	}
+
+	return results.ContinueProcessing()
+}
+
+// getApplicationSnapshot returns the all ApplicationSnapshots in the Application's namespace nil if it's not found.
+// In the case the List operation fails, an error will be returned.
+func (a *Adapter) getAllApplicationSnapshots() (*[]releasev1alpha1.ApplicationSnapshot, error) {
+	applicationSnapshots := &releasev1alpha1.ApplicationSnapshotList{}
+	opts := []client.ListOption{
+		client.InNamespace(a.application.Namespace),
+		client.MatchingFields{"spec.application": a.application.Name},
+	}
+
+	err := a.client.List(a.context, applicationSnapshots, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &applicationSnapshots.Items, nil
+}
+
+// compareApplicationSnapshots compares two ApplicationSnapshots and returns boolean true if their images match exactly
+func (a *Adapter) compareApplicationSnapshots(expectedApplicationSnapshot *releasev1alpha1.ApplicationSnapshot, foundApplicationSnapshot *releasev1alpha1.ApplicationSnapshot) bool {
+	allImagesMatch := true
```

Comment:
Maybe you can use reflect.DeepEqual for this. The current code would allow for the foundApplicationSnapshot to contain extra images. Not sure if this is something you can support.

Reply:
I haven't had good luck with using DeepEqual for this in testing, but I did add the check that the number of images matches.